routing/datasource.go

```diff
@@ -528,6 +531,69 @@ type routeTable struct {
 	created       time.Time
 }
 
+func fadeInState(now time.Time, duration time.Duration, detected time.Time) (time.Duration, bool) {
+	rel := now.Sub(detected)
+	return rel, rel > 0 && rel < duration
+}
+
+func fadeIn(now time.Time, duration time.Duration, exponent float64, detected time.Time) float64 {
+	rel, fadingIn := fadeInState(now, duration, detected)
+	if !fadingIn {
+		return 1
+	}
+
+	return math.Pow(float64(rel)/float64(duration), exponent)
+}
+
+func goodEndpoint(rnd *rand.Rand, ep *LBEndpoint, route *Route) bool {
+	now := time.Now()
+	f := fadeIn(
+		now,
+		route.LBFadeInDuration,
+		route.LBFadeInExponent,
+		ep.Detected,
+	)
+	goodFadeIn := f == 1 || rnd.Float64() < f
+
+	return goodFadeIn
+}
+
+func GoodEndpointerGoroutine(routes []*Route, quit <-chan struct{}, tick <-chan time.Time) {
+	rnd := rand.New(rand.NewSource(time.Now().UnixNano())) // #nosec
+
+	for {
+		select {
+		case <-quit:
+			return
+		case <-tick:
+			for _, ri := range routes {
+				if ri.Route.BackendType != eskip.LBBackend {
+					continue
+				}
+
+				if len(ri.Route.LBEndpoints) == 0 {
+					log.Errorf("failed to post-process LB route: %s, no endpoints defined", ri.Id)
+					continue
+				}
+
+				newGoodEndpoints := []LBEndpoint{}
+				newGoodHosts := map[string]struct{}{}
+				for i := range ri.LBEndpoints {
+					if goodEndpoint(rnd, &ri.LBEndpoints[i], ri) {
+						newGoodEndpoints = append(newGoodEndpoints, ri.LBEndpoints[i])
+						newGoodHosts[ri.LBEndpoints[i].Host] = struct{}{}
+					}
+				}
+
+				ri.Lock()
+				ri.GoodEndpoints = newGoodEndpoints
+				ri.GoodHosts = newGoodHosts
```

Comment:
Lets check if we really need it.

Reply:
In any case we do not need it for any algorithm other than consistentHash

Reply:
Maybe the intention is to have this for https://github.com/zalando/skipper/issues/2346

Reply:
That's an intention to make consistentHash.Apply to work in O(1). I fill up this map(set) here, but lookup in this map is O(1) regardless of its size.

Reply:
Not reviewed again the code but make sure the data structure is long living and only created at routing tree create/update